### PR TITLE
feat: image and json bidirectional channels integrated.

### PR DESCRIPTION
**Pull request by klepo**
- Prepare for version 0.7.0
- Channels class used to define channel data callbacks and contains send functions.
- TaskHandler abstract class removed.
- Comments, docs, type annotations, README.md updated.
- Era5gInterface exceptions.
- NetApp word removed.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -16,17 +20,25 @@ pip install era_5g_interface
...
+Channels class is used to define bidirectional channel data callbacks and contains send functions. It handles image 
+frames JPEG and H.264 encoding/decoding. The class cannot be used alone. The ServerChannels and ClientChannels 
+classes create callbacks and encoders/decoders.
+
+### ClientChannels and ServerChannels ([client_channels.py](era_5g_interface/client_channels.py), [server_channels.py](era_5g_interface/server_channels.py))
 
-### TaskHandler (task_handler.py)
+Client and server channels classes are used to define bidirectional channel (image ans json) callbacks and contains 
+send functions. They are used inside the NetAppClientBase in era_5g_client and NetworkApplicationServer in 
+era_5g_server. ServerChannels is used with Socketio Server object, ClientChannels is used with Socketio Client object.
```
Comment (ZdenekM) [inline README.md:33]: ...the same for package names

File: era_5g_interface/BUILD [hunk 2]
```diff
@@ -9,8 +9,8 @@ python_distribution(
     sdist=True,
     provides=setup_py(
         name="era_5g_interface",
-        version="0.6.1",
-        description="An interface for 5G-ERA NetApps",
+        version="0.7.0",
```
Comment (ZdenekM) [inline era_5g_interface/BUILD:12]: please set the version to `0.7.0rc0`

File: era_5g_interface/channels.py [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+import numpy as np
+import socketio
+
+from era_5g_interface.exceptions import UnknownChannelTypeUsed
+from era_5g_interface.h264_decoder import H264Decoder, H264DecoderError
+from era_5g_interface.h264_encoder import H264Encoder, H264EncoderError
+
+logger = logging.getLogger(__name__)
+
+DATA_NAMESPACE = str("/data")
```
Comment (ZdenekM) [inline era_5g_interface/channels.py:18]: Why `WHATEVER = str("whatever")`? What about using an enum?

File: era_5g_interface/channels.py [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+    JSON = 1
+    JPEG = 2
+    H264 = 3
+
+
+@dataclass
+class CallbackInfoClient:
+    """Callback info dataclass."""
+
+    type: ChannelType  # Channel type.
```
Comment (ZdenekM) [inline era_5g_interface/channels.py:39]: The comment is quite obvious :-)

File: era_5g_interface/channels.py [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+    callback: Callable[[dict], None]  # Custom callback.
+    error_event: str = DATA_ERROR_EVENT  # Custom error event name.
+
+
+@dataclass
+class CallbackInfoServer:
```
Comment (ZdenekM) [inline era_5g_interface/channels.py:45]: What is the point of having two exactly the same classes?

File: era_5g_interface/channels.py [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+
+
+class Channels(ABC):
+    """Channels class is used to define channel data callbacks and contains send functions.
+
+    It handles image frames JPEG and H.264 encoding/decoding. The class cannot be used alone. The ServerChannels and
+    ClientChannels classes create callbacks and encoders/decoders.
+    """
+
+    _callbacks_info: Union[Dict[str, CallbackInfoClient], Dict[str, CallbackInfoServer]]
```
Comment (ZdenekM) [inline era_5g_interface/channels.py:60]: Why is this "static"?

File: era_5g_interface/channels.py [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+
+        if channel_type is ChannelType.H264:
+            if eio_sid not in self._encoders:
+                self._encoders[eio_sid] = H264Encoder(frame.shape[1], frame.shape[0], options=encoding_options)
+
+        try:
+            is_key_frame = True
+            if channel_type is ChannelType.H264:
+                frame_encoded = self._encoders[eio_sid].encode_ndarray(frame)
+                data["h264"] = True
```
Comment (ZdenekM) [inline era_5g_interface/channels.py:144]: What about dataclass for this data?

File: era_5g_interface/channels.py [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+    ) -> None:
+        """Send general JSON data.
+
+        Args:
+            data (Dict[str, Any]): JSON data.
+            event (str): Event name.
+            sid (str, optional): Namespace sid - mandatory when sending from the server side to the client.
+            can_be_dropped (bool): If data can be lost due to back pressure.
+        """
+        if type(self._sio) is socketio.Client:
```
Comment (ZdenekM) [inline era_5g_interface/channels.py:178]: Use rather `isinstance`.

File: era_5g_interface/dataclasses/control_command.py [hunk 9]
```diff
@@ -48,4 +48,4 @@ class ControlCommand:
 
     cmd_type: ControlCmdType
     clear_queue: bool = False
-    data: dict = field(default_factory=dict)
+    data: Optional[dict] = field(default_factory=dict)
```
Comment (ZdenekM) [inline era_5g_interface/dataclasses/control_command.py:51]: We should use `Dict` as a type hint to be compatible with Python 3.8.

File: era_5g_interface/channels.py [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+                    },
+                    self._callbacks_info[event].error_event,
+                    sid=sid,
+                )
+                return None
+            self._decoders[eio_sid].last_timestamp = timestamp
+
+        try:
+            if eio_sid in self._decoders:
+                frame_decoded = self._decoders[eio_sid].decode_packet_data(data["frame"])
```
Comment (ZdenekM) [inline era_5g_interface/channels.py:258]: I would use `H264DecoderError` here

File: era_5g_interface/channels.py [hunk 11]
```diff
@@ -0,0 +1,294 @@
...
+            else:
+                frame_decoded = cv2.imdecode(np.frombuffer(data["frame"], dtype=np.uint8), cv2.IMREAD_COLOR)
```
Comment (ZdenekM) [inline era_5g_interface/channels.py:260]: ...and some CV2 exception (or whatever it raises) here

File: era_5g_interface/interface_helpers.py [hunk 12]
```diff
@@ -17,35 +17,67 @@
...
+        """
+
+        self.retries = Retry(total=0, read=0, connect=0, backoff_factor=0, status_forcelist=[429, 500, 502, 503, 504])
         self.session: requests.Session = requests.Session()
         self.adapter: HTTPAdapter = HTTPAdapter(max_retries=self.retries)
         self.session.mount(MIDDLEWARE_ADDRESS, self.adapter)
         self.connection_error = False
 
-    def send_middleware_heart_beat_request(self, headers: dict, json: dict) -> None:
+    def _send_middleware_heart_beat_request(self, headers: dict, json: dict) -> None:
```
Comment (ZdenekM) [inline era_5g_interface/interface_helpers.py:73]: Should be `Dict`

File: era_5g_interface/channels.py (hunk 5 above)
Comment (klepo) [reply era_5g_interface/channels.py:45]: Due to different callback type annotations and exclusive use in individual classes.

File: era_5g_interface/channels.py (hunk 7 above)
Comment (klepo) [reply era_5g_interface/channels.py:144]: Good idea.. TODO

File: era_5g_interface/channels.py (hunk 5 above)
Comment (ZdenekM) [reply era_5g_interface/channels.py:45]: I see, I missed that the annotation is different.

File: era_5g_interface/channels.py (hunk 10 above)
Comment (klepo) [reply era_5g_interface/channels.py:258]: I don't understand where you mean.

File: era_5g_interface/channels.py (hunk 11 above)
Comment (klepo) [reply era_5g_interface/channels.py:260]: Again, I don't understand.

File: era_5g_interface/channels.py (hunk 10 above)
Comment (klepo) [reply era_5g_interface/channels.py:261]: OK

File: era_5g_interface/channels.py (hunk 6 above)
Comment (ZdenekM) [reply era_5g_interface/channels.py:62]: Ok, I see. In such cases, an explaining comment might be helpful.

**Review by ZdenekM — APPROVED** [review]

**Pull request merged**